### DISCUSSION TITLE
ESP32-S3 octal PSRAM not visible

### DISCUSSION THREAD

**Discussion by brokenbrigde9000** [micropython/micropython#12760] [ESP32]
I got myself some ESP32-S3-WROOM-1U-N16R8 units and expected to be able to utilize more RAM than the 'regular' non-R versions. Maybe I don't fully understand how these things work or I'm doing something wrong, but I am not able to use/see more than ~250kB of memory. I have flashed the units with v1.21.0 (2023-10-05)  with support for Octal-SPIRAM but to no avail.

How can I use the memory on these devices? I want to process some long lists which currently cause me to run out of memory.

**Comment by straga** [comment]
Check again what you flashed.


<img width="583" alt="image" src="https://github.com/micropython/micropython/assets/34532/9f580f90-ff3a-4e9f-9e12-8c0162bbb21a">

![image](https://github.com/micropython/micropython/assets/34532/58b1fa78-8b49-43e7-96fd-8e650409866d)

**Reply by brokenbrigde9000** [reply]
I have this:

MicroPython v1.21.0 on 2023-10-06; Generic ESP32S3 module with Octal-SPIRAM with ESP32S3

<img width="591" alt="memtest" src="https://github.com/micropython/micropython/assets/59135284/82e94b0a-9929-4eb6-8839-eb973c8ef45f">

**Comment by bixb922** [comment]
I just flashed a ESP32-S3-WROOM-1 N16R8 with a recent preview and see the 8Mb:
```
MicroPython v1.22.0-preview.9.g5f0bd33b7 on 2023-10-13; Generic ESP32S3 module with Octal-SPIRAM with ESP32S3
Type "help()" for more information.
>>> import micropython
>>> micropython.mem_info()
stack: 704 out of 15360
GC: total: 64000, used: 13360, free: 50640, max new split: 8257536
 No. of 1-blocks: 196, 2-blocks: 32, max blk sz: 37, max free sz: 3152
>>> import gc
>>> gc.mem_free()
8307632
>>> 
```
My ESP32-S3 seems to be the same model as on your photo. There have been some nice enhancements to the SPIRAM support very recently. Try with the last preview available.

To make the 16 Mb of flash memory visible, there is a nice tool: mp-image-tool-esp32 at #12751

**Reply by brokenbrigde9000** [reply]
I flashed MicroPython v1.22.0-preview.51.g91a3f1839 (generic ESP32S3 module with octal-SPIRAM) but no sigar. When quering device info in Thonny the 16MB of flash memory is mentioned,  but this not the most important to me. I'm more interested in that 8mb RAM that does show in your mem_info() but still is missing in mine.  I only see 242432 bytes.

Can connecting to certain pins cause problems? I see pin 36 and 37 might have something to do with SPI and they are connected with a transistor and pulldown resistor.

**Reply by sophiedegran** [reply]
I suppose you try in the right way.. So if the psram is not well recognize:
 
- perharps it's a Quad SPI (QD) and not Octal SPI (OT).. i've heard of that but normally QD is for 2MB max on ESP32S3.. So you flash is OT and psram QD .. try another image for that ☺

**Reply by bixb922** [reply]
Fwiw: On my esp32-s3, pins 33 to 38 are not available with octal SPIRAM. Some esp32-s3 docs state:
for flash/RAM octal mode GPIO 33, 34, 35, 36, 37, 38 are not available.
Idk if that can have the symptoms you describe

**Reply by brokenbrigde9000** [reply]
Thats too bad, that wasn't in the datasheet i viewed. I will try to disconnect these traces from the ESP32 and see if the SPIRAM is initiated properly.

**Comment by bixb922** [comment]
Can you post the schematics?

**Comment by rkompass** [comment]
Could it be that you observe the behaviour described [here](https://github.com/micropython/micropython/issues/12253)?
TLDR: The available RAM now on ESP32 grows on use and is reported to be smaller in the beginning.

**Comment by jimmo** [comment]
@brokenbrigde9000 I think @rkompass is right -- can you take a look at https://github.com/orgs/micropython/discussions/12316 and see if that explains what you're seeing. (In which case, working as intended, the heap will grow automatically)

**Comment by brokenbrigde9000** [comment]
The problem is solved, I indeed made a mistake connecting something to pin 35, 36 and 37. I didn't see that immediately because I never actually called machine.Pin on these three pins. I would have known because you actually can't use those pins in the octal spiram firmware version. I cut those lines and immediately the 8mb of memory became available. I can now temporarily store an enormous amount of data! Thanks for the help.
